homeassistant/core.py

```diff
@@ -1907,18 +1932,27 @@ async def async_call(
             raise asyncio.CancelledError
         if task.done():
             # Propagate any exceptions that might have happened during service call.
-            task.result()
+            result = task.result()
             # Service call completed successfully!
+            if return_values:
+                if not isinstance(result, dict):
+                    raise HomeAssistantError(
+                        f"Service return value expected a dictionary, was {type(result)}"
+                    )
+                return result
             return True
+
         # Service call task did not complete before timeout expired.
         # Let it keep running in background.
         self._run_service_in_background(task, service_call)
```

Comment:
We should not run in the background if the service was requested to return a result, right?

Reply:
Handled in other PR to remove limits.